python/lib/modeldata/dmod/modeldata/hydrofabric/geopackage_hydrofabric.py

```diff
@@ -0,0 +1,593 @@
+        subset_query_setups: Dict[str, Tuple[str, Callable[[], Iterable[str]]]] = {
+            'flowpaths': ('realized_catchment', lambda: subset.catchment_ids),
+            'divides': ('id', lambda: subset.catchment_ids),
+            'nexus': ('id', lambda: subset.nexus_ids),
+            'flowpath_attributes': ('id', lambda: new_dfs['flowpaths']['id']),
+            'flowpath_edge_list': ('id', lambda: new_dfs['flowpaths']['id']),
+            'crosswalk': ('id', lambda: new_dfs['flowpaths']['id']),
+            'cfe_noahowp_attributes': ('id', lambda: subset.catchment_ids),
+            'forcing_metadata': ('id', lambda: subset.catchment_ids)
+        }
+
+        # Then, apply this logic to every encountered layer to create subset layer/dataframe to use to init new instance
+        def subset_layer(layer_name: str):
+            dataframe = self._dataframes[layer_name]
+            id_search_col = subset_query_setups[layer_name][0]
+            applicable_ids = subset_query_setups[layer_name][1]()
+            new_dfs[layer_name] = dataframe.loc[dataframe[id_search_col].isin(applicable_ids)]
+
+        # Subset 'flowpaths' layer first; it's ids may be needed for subsetting other things like 'flowpath_edge_list'
+        if 'flowpaths' in self._layer_names:
+            subset_layer('flowpaths')
+
+        # Now, generate the rest of the subset layers/dataframes
+        for layer in [ln for ln in self._layer_names if ln != 'flowpaths']:
+            subset_layer(layer)
+
+        return GeoPackageHydrofabric(layer_names=self._layer_names, layer_dataframes=new_dfs)
+
+    def is_catchment_recognized(self, catchment_id: str) -> bool:
+        """
+        Test whether a catchment is recognized.
+
+        Parameters
+        ----------
+        catchment_id : str
+            The id of the catchment.
+
+        Returns
+        -------
+        bool
+            Whether the catchment is recognized.
+        """
+        return catchment_id in self._dataframes[self._DIVIDES_LAYER_NAME][self._DIVIDES_CAT_ID_COL].values
+
+    @property
+    def is_conus(self) -> bool:
+        """
+        Whether this hydrofabric represents all of CONUS.
+
+        Returns
+        -------
+        bool
+            Whether this hydrofabric represents all of CONUS.
+        """
+        return self._is_conus
+
+    def is_nexus_recognized(self, nexus_id: str) -> bool:
+        """
+       Test whether a nexus is recognized.
+
+       Parameters
+       ----------
+       nexus_id : str
+           The id of the nexus.
+
+       Returns
+       -------
+       bool
+           Whether the nexus is recognized.
+       """
+        return nexus_id in self._dataframes[self._NEXUS_LAYER_NAME][self._NEXUS_NEX_ID_COL].values
+
+    @property
+    def roots(self) -> FrozenSet[str]:
+        """
+        Get the ids of the root nodes of the hydrofabric graph.
```

Comment:
Can you add a little bit to what you mean by _root_ nodes? To my knowledge this isnt in the hydrofabric vernacular (although, im likely wrong).

Reply:
These are vertices that do not have an incoming edge in the hydrofabric's directed acyclic graph representation; i.e., the features from which you cannot travel further upstream (which, as I understand things, should always be catchment type features).

The term "root" is perhaps slightly misused in a strict graph theory context, at least in general.  It is applicable for at least some hydrofabric graphs, which may be how it caught on.  I don't think this usage originated or is exclusive to this class.